## Part 1: Import Networks from Statoil Files
This example explains how to use the OpenPNM.Utilies.IO.Statoil class to import a network produced by the Maximal Ball network extraction code developed by Martin Blunt's group at Imperial College London. The code is available from him upon request, but they offer a small library of pre-extracted networks on their [website](https://www.imperial.ac.uk/engineering/departments/earth-science/research/research-groups/perm/research/pore-scale-modelling/micro-ct-images-and-networks/).

In [1]:
import scipy as sp
import openpnm as op

C:\Users\Jeff\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


The following assumes that the folder containing the 'dat' files is in a directory called 'fixtures' in the same directory as this script.  You can also enter a full path to the files.

In [2]:
from pathlib import Path
path = Path('../fixtures/ICL-Sandstone(Berea)/')
project = op.io.Statoil.load(path=path, prefix='Berea')
pn = project.network
pn.name = 'berea'

This import class extracts all the information contained in the 'Statoil' files, such as sizes, locations and connectivity. Note that the ``io`` classes return a ``project`` object, and the network itself can be accessed using the ``network`` attribute.

At this point, the network can be visualized in Paraview. A suitable '.vtp' file can be created with:

In [3]:
project.export_data(network=pn, filename='imported_statoil')

The resulting network is shown below:

![](http://i.imgur.com/771T36M.png)

### Clean up network topology

Although it's not clear in the network image, there are a number of isolated and disconnected pores in the network.  These are either naturally part of the sandstone, or artifacts of the Maximal Ball algorithm.  In any event, these must be removed before proceeding since they cause problems for the matrix solver.  The easiest way to find these is to use the ```check_network_health``` function on the network object.  This will return a dictionary with several key attributes, including a list of which pores are isolated. These can then be trimmed using the ``trim`` function in the ``topotools`` module.

In [4]:
print('Number of pores before trimming: ', pn.Np)
h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])
print('Number of pores after trimming: ', pn.Np)

Number of pores before trimming:  6298
Number of pores after trimming:  6004


### Dealing with Inlet and Outlet Pores
When importing Statoil networks, OpenPNM must perform some 'optimizations' to make the network compatible.  The main problem is that the original network contains a large number of throats connecting actual internal pores to fictitious 'reservoir' pores.  OpenPNM strips away all these throats since 'headless throats' break the graph theory representation.  OpenPNM then labels the real internal pores as either 'inlet' or 'outlet' if they were connected to one of these fictitious reservoirs.  

It is fairly simple to add a new pores to each end of the domain and stitch tehm to the internal pores labelled 'inlet' and 'outlet', but this introduces a subsequent complication that the new pores don't have any geometry properties.  For this example, we will not add boundary pores, but just the pores on the inlet and outlet faces.  

## Part 2: Calculating Permeability of the Network

### Setup Geometry, Phase, and Physics Objects

In OpenPNM 2+ it is optional to define Geometry and Physics objects (These are really only necessary for simulations with diverse geometrical properties in different regions, resulting in different physical processes in each region, such as multiscale networks for instance).  It is still necessary to define **Phase** objects:

In [5]:
water = op.phases.Water(network=pn)

Also, we must add a few geometrical properties that were not included in the extracted network:

In [6]:
pn['pore.diameter'] = 2*pn['pore.radius']
pn['throat.diameter'] = 2*pn['throat.radius']
pn.add_model(propname='pore.area',
             model=op.models.geometry.pore_area.sphere)
pn.add_model(propname='throat.area',
             model=op.models.geometry.throat_area.cylinder)

### Apply Pore-Scale Models

We must add the hagen-poiseuille model for calculating the conductance.  In OpenPNM 2+ it is possible to add Physics models to Phase objects, which is often simpler than than applying the same model to multiple Physics.

In [7]:
water.add_model(propname='throat.hydraulic_conductance',
                model=op.models.physics.hydraulic_conductance.hagen_poiseuille)

Recall that boundary pores and throats had no geometrical properties associated with them, so the   hydraulic conductances of boundary throats will be undefined (filled with NaNs):


In [8]:
print(water['throat.hydraulic_conductance'])

[3.11794440e-13 1.18780711e-12 2.44420114e-13 ... 2.80503466e-11
 2.16232529e-11 3.70909786e-11]


### Run StokesFlow Algorithm
Finally, we can create a **StokesFlow** object to run some fluid flow simulations:

In [9]:
flow = op.algorithms.StokesFlow(network=pn, phase=water)
flow.set_dirichlet_BC(pores=pn.pores('inlets'), values=200000)
flow.set_dirichlet_BC(pores=pn.pores('outlets'), values=100000)
flow.run()

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Running ReactiveTransport
Tolerance not met: 142387023616295.0
Solution converged: 0.0


The resulting pressure field can be visualized in Paraview, giving the following:

![](https://i.imgur.com/AIK6FbJ.png)

### Determination of Permeability Coefficient

The way to calculate K is the determine each of the values in Darcy's law manually and solve for K, such that $$ K = \frac{Q\mu L} {\Delta P A} $$

In [10]:
# Get the average value of the fluid viscosity
mu = sp.mean(water['pore.viscosity'])
# Specify a pressure difference (in Pa)
delta_P = 100000
# Using the rate method of the StokesFlow algorithm
Q = sp.absolute(flow.rate(pores=pn.pores('inlets')))
# Because we know the inlets and outlets are at x=0 and x=X
Lx = sp.amax(pn['pore.coords'][:, 0]) - sp.amin(pn['pore.coords'][:, 0])
A = Lx*Lx  # Since the network is cubic Lx = Ly = Lz
K = Q*mu*Lx/(delta_P*A)
print(K)

[1.96517979e-12]


Using this approach the K value is 1.97 mD, which is very close to the value of 1.1 mD reported by the Imperial Team on their network.  The difference is due to the different models used to calculate hydraulic conductance.